File: Components/DevExpressVCL/Demos/ExpressPivotGrid/CBuilder/MultipleTotalsDemo/MultipleTotalsMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "MultipleTotalsMain.h"
//---------------------------------------------------------------------------
#include "DemoBasicDM.h"
#pragma package(smart_init)
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxDBPivotGrid"
#pragma resource "*.dfm"
TfrmMultipleTotals *frmMultipleTotals;
//---------------------------------------------------------------------------
__fastcall TfrmMultipleTotals::TfrmMultipleTotals(TComponent* Owner)
        : TfrmDemoBasicMain(Owner)
{
}

void __fastcall TfrmMultipleTotals::FormCreate(TObject* Sender)
{
  TfrmDemoBasicMain::FormCreate(Sender); 
  PivotGrid()->ApplyBestFit();
}

TcxCustomPivotGrid* __fastcall TfrmMultipleTotals::PivotGrid()
{
  return DBPivotGrid;
}

//---------------------------------------------------------------------------

double VarToDoubleEx(Variant AValue)
{
  if (!VarIsNull(AValue)) {
    return ((double)AValue);
  }
  else {
    return 0;
  }
}

void __fastcall TfrmMultipleTotals::CalculateCustomSummary(
      TcxPivotGridField *Sender, TcxPivotGridCrossCellSummary *ASummary)
{
  TcxPivotGridGroupItem *ARow;
  TcxPivotGridGroupItem *AColumn;
  if (pgfCompanyName->Area == faRow) {
    ARow = ASummary->Owner->Row;
    AColumn = ASummary->Owner->Column;
  }
  else {
    ARow = ASummary->Owner->Column;
    AColumn = ASummary->Owner->Row;
  }
  if (AColumn->Parent == NULL){
    ASummary->Custom = 100;
  }
  else{
    double ASum = VarToDoubleEx(ASummary->Sum);
    double AGrandTotalValue = VarToDoubleEx(ARow->GetCellByCrossItem(
      AColumn->Parent)->GetSummaryByField(Sender, stSum));
    if (AGrandTotalValue != 0) {
      ASummary->Custom = (ASum / AGrandTotalValue * 100);
    }
    else {
      ASummary->Custom = NULL;
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmMultipleTotals::PivotGridStylesGetGroupHeaderStyle(
      TcxCustomPivotGrid *Sender, TcxPivotGridViewDataItem *AItem,
      TcxStyle *&AStyle)
{
   if (AItem->IsTotal || !AItem->Expanded & (AItem->Field == pgfCompanyName)) {
     AStyle = dmOrders->stBoldBlueFont;
   }
}
//---------------------------------------------------------------------------

void __fastcall TfrmMultipleTotals::PivotGridStylesGetContentStyle(
      TcxCustomPivotGrid *Sender, TcxPivotGridDataCellViewInfo *ACell,
      TcxStyle *&AStyle)
{
  if (ACell->SummaryType == stCustom) {
    AStyle = dmOrders->stBoldRedFont;
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressPivotGrid/Delphi/MultipleTotalsDemo/MultipleTotalsMain.pas
unit MultipleTotalsMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, cxLookAndFeels, Menus, StdCtrls, cxControls,
  cxCustomPivotGrid, cxDBPivotGrid, DemoBasicDM, cxStyles, cxClasses,
  cxGraphics, cxCustomData, Variants;

type
  TfrmMultipleTotals = class(TfrmDemoBasicMain)
    DBPivotGrid: TcxDBPivotGrid;
    pgfPurchaseMonth: TcxDBPivotGridField;
    cxDBPivotGrid1PaymentType: TcxDBPivotGridField;
    cxDBPivotGrid1Quantity: TcxDBPivotGridField;
    cxDBPivotGrid1CarName: TcxDBPivotGridField;
    cxDBPivotGrid1UnitPrice: TcxDBPivotGridField;
    pgfCompanyName: TcxDBPivotGridField;
    cxDBPivotGrid1PaymentAmount: TcxDBPivotGridField;
    pgfPurchaseQuarter: TcxDBPivotGridField;
    procedure PivotGridStylesGetGroupHeaderStyle(Sender: TcxCustomPivotGrid;
      AItem: TcxPivotGridViewDataItem; var AStyle: TcxStyle);
    procedure CalculateCustomSummary(
      Sender: TcxPivotGridField; ASummary: TcxPivotGridCrossCellSummary);
    procedure PivotGridStylesGetContentStyle(Sender: TcxCustomPivotGrid;
      ACell: TcxPivotGridDataCellViewInfo; var AStyle: TcxStyle);
  private
  protected
    function GetPivotGrid: TcxCustomPivotGrid; override;
  public
    { Public declarations }
  end;

var
  frmMultipleTotals: TfrmMultipleTotals;

implementation

{$R *.dfm}

function TfrmMultipleTotals.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := DBPivotGrid;
end;

procedure TfrmMultipleTotals.PivotGridStylesGetGroupHeaderStyle(
  Sender: TcxCustomPivotGrid; AItem: TcxPivotGridViewDataItem;
  var AStyle: TcxStyle);
begin
   if AItem.IsTotal or not AItem.Expanded and (AItem.Field = pgfCompanyName) then
     AStyle := dmOrders.stBoldBlueFont;
end;

procedure TfrmMultipleTotals.PivotGridStylesGetContentStyle(
  Sender: TcxCustomPivotGrid; ACell: TcxPivotGridDataCellViewInfo;
  var AStyle: TcxStyle);
begin
  if ACell.SummaryType = stCustom then
    AStyle := dmOrders.stBoldRedFont;
end;

procedure TfrmMultipleTotals.CalculateCustomSummary(
  Sender: TcxPivotGridField; ASummary: TcxPivotGridCrossCellSummary);

  function VarToDouble(const AValue: Variant): Double;
  begin
    Result := 0;
    if not VarIsNull(AValue) then
      Result := AValue;
  end;

var
  ARow, AColumn: TcxPivotGridGroupItem;
  AGrandTotalValue, ASum: Double;

begin
  if pgfCompanyName.Area = faRow then
  begin
    ARow := ASummary.Owner.Row;
    AColumn := ASummary.Owner.Column;
  end
  else // rotated to 90 degrees
  begin
    ARow := ASummary.Owner.Column;
    AColumn := ASummary.Owner.Row;
  end;
  if AColumn.Parent = nil then
    ASummary.Custom := 100
  else
  begin
    AGrandTotalValue := VarToDouble(ARow.GetCellByCrossItem(
      AColumn.Parent).GetSummaryByField(Sender, stSum));
    ASum := VarToDouble(ASummary.Sum);
    if AGrandTotalValue <> 0 then
      ASummary.Custom := ASum / AGrandTotalValue  * 100
    else
      ASummary.Custom := 0;
  end;
end;



end.
